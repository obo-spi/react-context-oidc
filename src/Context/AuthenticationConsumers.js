import React, { Fragment } from 'react';
import { withRouter } from 'react-router-dom';
import {
  compose,
  branch,
  lifecycle,
  defaultProps,
  renderComponent
} from 'recompose';

import { withOidcUser } from './AuthenticationContext.container';
import {
  authenticateUser,
  getUserManager,
  oidcLog,
  isRequireAuthentication
} from '../Services';
import { Authenticating } from '../OidcComponents';

const lifecycleComponent = {
  async componentDidMount() {
    oidcLog.info('Protected component mounted');
    const usermanager = getUserManager();
    await authenticateUser(usermanager, this.props.location)();
  }
};

const wrapAuthenticating = () => () => <Authenticating />;

const Dummy = props => {
  return <Fragment>{props.children}</Fragment>;
};

const withDefaultProps = defaultProps({
  isEnabled: true
});

export const withOidcSecure = compose(
  withDefaultProps,
  branch(props => !props.isEnabled, renderComponent(Dummy)),
  withOidcUser,
  withRouter,
  lifecycle(lifecycleComponent),
  branch(isRequireAuthentication(), wrapAuthenticating)
);

const OidcSecure = props => {
  const Secure = withOidcSecure(Dummy);
  return <Secure {...props}>{props.children}</Secure>;
};

export default OidcSecure;
