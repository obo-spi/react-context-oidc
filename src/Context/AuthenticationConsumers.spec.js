import React from 'react';
import { renderHook, cleanup } from 'react-hooks-testing-library';
import { render, waitForElement } from 'react-testing-library';
import { AuthenticationContext } from './AuthenticationContextCreator';
import 'jest-dom/extend-expect';

import * as consumers from './AuthenticationConsumers';

const getWrapper = values => ({ children }) => (
  <AuthenticationContext.Provider value={values}>
    {children}
  </AuthenticationContext.Provider>
);
describe('Consumers service tests suite', () => {
  const values = {
    isEnabled: true,
    oidcUser: { user: 'tom' },
    isLogout: false
  };
  const getUserManager = () => ({
    param: 'Mock version'
  });
  const authenticateUser = jest.fn(() => jest.fn());

  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterEach(cleanup);

  it('should not call authentication when enabled', () => {
    renderHook(
      () =>
        consumers.useOidcSecureWithService(
          authenticateUser,
          getUserManager,
          '/locationUser'
        ),
      { wrapper: getWrapper(values) }
    );
    expect(authenticateUser).toBeCalledWith(
      { param: 'Mock version' },
      '/locationUser'
    );
  });

  it('should NOT call authentication when disabled', () => {
    renderHook(
      () =>
        consumers.useOidcSecureWithService(
          authenticateUser,
          getUserManager,
          '/locationUser'
        ),
      { wrapper: getWrapper({ ...values, isEnabled: false }) }
    ),
      expect(authenticateUser).not.toBeCalled();
  });

  it('should NOT call authentication if user logout', () => {
    renderHook(
      () =>
        consumers.useOidcSecureWithService(
          authenticateUser,
          getUserManager,
          '/locationUser'
        ),
      { wrapper: getWrapper({ ...values, isLogout: true }) }
    ),
      expect(authenticateUser).not.toBeCalled();
  });

  it('withOidcUser should inject oidcUser if user connected', () => {
    const testingComponent = ({ oidcUser }) => (
      <div>
        <span data-testid="username">{oidcUser}</span>
      </div>
    );
    const Component = consumers.withOidcUser(testingComponent);
    const { getByTestId, asFragment } = render(<Component />, {
      wrapper: getWrapper({ oidcUser: 'Franck' })
    });
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId('username')).toHaveTextContent('Franck');
  });

  it('withOidcSecurewithRouter should call authentication and AuthenticatingComponent during authentication process', () => {
    const contextMock = {
      isEnabled: true,
      oidcUser: null,
      isLogout: false
    };

    const mockedProps = {
      location: '/myurl',
      authenticateUser,
      getUserManager
    };

    const testingComponent = () => <div data-testid="componentMount" />;
    const Component = consumers.withOidcSecurewithRouter(testingComponent);
    const { queryByTestId, asFragment } = render(
      <Component {...mockedProps} />,
      {
        wrapper: getWrapper({ ...contextMock })
      }
    );
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('componentMount')).toBeNull();
  });

  it('withOidcSecurewithRouter should show component if user connected', () => {
    const contextMock = {
      isEnabled: true,
      oidcUser: { user: 'tom' },
      isLogout: false
    };

    const mockedProps = {
      location: '/myurl',
      authenticateUser,
      getUserManager
    };

    const testingComponent = () => <div data-testid="componentMount" />;
    const Component = consumers.withOidcSecurewithRouter(testingComponent);
    const { queryByTestId } = render(<Component {...mockedProps} />, {
      wrapper: getWrapper({ ...contextMock })
    });
    const greetingNode = queryByTestId('componentMount');
    expect(queryByTestId('componentMount')).not.toBeNull();
  });
});
