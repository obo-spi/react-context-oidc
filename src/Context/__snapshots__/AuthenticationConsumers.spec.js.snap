// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consumers service tests suite withOidcSecurewithRouter should call authentication and AuthenticatingComponent during authentication process 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
    >
      <h1>
        Authentification en cours
      </h1>
      <p>
        Vous allez être redirigé sur la page de login
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Consumers service tests suite withOidcUser should inject oidcUser if user connected 1`] = `
<DocumentFragment>
  <div>
    <span
      data-testid="username"
    >
      Franck
    </span>
  </div>
</DocumentFragment>
`;
